lxd/instance/drivers/driver_qemu.go

```diff
@@ -7450,74 +7460,101 @@ func (d *qemu) getVsockID() (uint32, error) {
-			// Skip the vsockID if another syscall error was encountered.
-			return false
+	// 0x4008AF60 = VHOST_VSOCK_SET_GUEST_CID = _IOW(VHOST_VIRTIO, 0x60, __u64)
+	_, _, errno := unix.Syscall(unix.SYS_IOCTL, f.Fd(), 0x4008AF60, uintptr(unsafe.Pointer(&vsockID)))
+	if errno != 0 {
+		if !errors.Is(errno, unix.EADDRINUSE) {
+			return nil, fmt.Errorf("Failed ioctl syscall to vhost socket: %q", errno.Error())
 		}
 
-		// The syscall to the vsock device returned "no such device".
-		// This means the address (Context ID) is free.
-		return true
+		// vsock Context ID is already in use.
+		return nil, nil
 	}
 
-	// Address is already in use.
-	c.Close()
-	return false
+	revert.Success()
+	return f, nil
+}
+
+// acquireExistingVsockID tries to acquire an already existing vsock Context ID from volatile.
+// It returns both the acquired ID and opened vsock file handle for QEMU.
+func (d *qemu) acquireExistingVsockID() (uint32, *os.File, error) {
+	vsockID, err := d.getVsockID()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	// Check if the vsockID from last VM start is still not acquired in case the VM was stopped.
+	f, err := d.acquireVsockID(vsockID)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	return vsockID, f, nil
 }
 
-// nextVsockID returns the next free vsock Context ID for the VM.
-// It tries to acquire one randomly until the timeout exceeds.
-func (d *qemu) nextVsockID() (uint32, error) {
+// nextVsockID tries to acquire the next free vsock Context ID for the VM.
+// It returns both the acquired ID and opened vsock file handle for QEMU.
+func (d *qemu) nextVsockID() (uint32, *os.File, error) {
 	// Check if vsock ID from last VM start is present in volatile, then use that.
 	// This allows a running VM to be recovered after DB record deletion and that an agent connection still works
 	// after the VM's instance ID has changed.
 	// Continue in case of error since the caller requires a valid vsockID in any case.
-	vsockID, err := d.getVsockID()
-	if err == nil {
-		// Check if the vsock ID from last VM start is still not acquired in case the VM was stopped.
-		if d.freeVsockID(vsockID) {
-			return vsockID, nil
-		}
+	vsockID, vsockF, _ := d.acquireExistingVsockID()
+	if vsockID != 0 && vsockF != nil {
+		return vsockID, vsockF, nil
 	}
 
+	// Ignore the error from before and start to acquire a new Context ID.
 	instanceUUID := uuid.Parse(d.localConfig["volatile.uuid"])
 	if instanceUUID == nil {
-		return 0, fmt.Errorf("Failed to parse instance UUID from volatile.uuid")
+		return 0, nil, fmt.Errorf("Failed to parse instance UUID from volatile.uuid")
 	}
 
 	r, err := util.GetStableRandomGenerator(instanceUUID.String())
 	if err != nil {
-		return 0, fmt.Errorf("Failed generating stable random seed from instance UUID %q: %w", instanceUUID, err)
+		return 0, nil, fmt.Errorf("Failed generating stable random seed from instance UUID %q: %w", instanceUUID, err)
 	}
 
-	timeout := 5 * time.Second
+	timer := time.NewTimer(5 * time.Second)
```

Comment:
This creates a go routine, which feels unnecessary in this situation (as we are not waiting on another async function to return).

Suggest using time.After such as in ZFS' `deactivateVolume`.

Reply:
I switched to using this pattern to allow for having a happy path (not the final error return at the bottom of the func).

Reply:
Changed, still happy path, but new pattern :)